plugin/plugins/filesystem/nodes/ipfs.go

```diff
@@ -0,0 +1,222 @@
+
+	if shouldClone(names) {
+		id.Logger.Debugf("Walk cloned")
+		return []p9.QID{id.Qid}, id, nil
+	}
+
+	qids := make([]p9.QID, 0, len(names))
+
+	walkedNode := &IPFS{} // [walk(5)] id represents fid, walkedNode represents newfid
+	*walkedNode = *id
+
+	var err error
+	for _, name := range names {
+		//TODO: timerctx; don't want to hang forever
+		if walkedNode.Path, err = id.core.ResolvePath(id.Ctx, corepath.Join(walkedNode.Path, name)); err != nil {
+			return qids, nil, err
+		}
+
+		ipldNode, err := id.core.Dag().Get(id.Ctx, walkedNode.Path.Cid())
+		if err != nil {
+			return qids, nil, err
+		}
+
+		//TODO: this is too opague; we want core path => qid, dirent isn't necessary
+		dirEnt := &p9.Dirent{}
+		if err = ipldStat(dirEnt, ipldNode); err != nil {
+			return qids, nil, err
+		}
+
+		walkedNode.Qid = dirEnt.QID
+		//
+		qids = append(qids, walkedNode.Qid)
+	}
+
+	id.Logger.Debugf("Walk ret %v, %v", qids, walkedNode)
+	return qids, walkedNode, err
+}
+
+func (id *IPFS) Open(mode p9.OpenFlags) (p9.QID, uint32, error) {
+	id.Logger.Debugf("Open %q", id.Path)
+
+	// set up  handle amenities
+	var handleContext context.Context
+	handleContext, cancel := context.WithCancel(id.Ctx)
+	id.cancel = cancel
+
+	// handle directories
+	if id.meta.Mode.IsDir() {
+		c, err := id.core.Unixfs().Ls(handleContext, id.Path)
+		if err != nil {
+			id.Logger.Errorf("hit\n%q\n%#v\n", id.Path, err)
+			cancel()
+			return id.Qid, 0, err
+		}
+
+		id.directory = &directoryStream{
+			entryChan: c,
+		}
+		return id.Qid, 0, nil
+	}
+
+	// handle files
+	apiNode, err := id.core.Unixfs().Get(handleContext, id.Path)
+	if err != nil {
+		id.Logger.Errorf("hit\n%q\n%#v\n", id.Path, err)
+		cancel()
+		return id.Qid, 0, err
+	}
+
+	var ok bool
+	if id.file, ok = apiNode.(files.File); !ok {
+		cancel()
+		return id.Qid, 0, fmt.Errorf("%q does not appear to be a file: %T", id.Path.String(), apiNode)
+	}
+
+	return id.Qid, ipfsBlockSize, nil
+}
+
+func (id *IPFS) Readdir(offset uint64, count uint32) ([]p9.Dirent, error) {
+	id.Logger.Debugf("Readdir %d %d", offset, count)
+
+	if id.directory == nil {
+		return nil, fmt.Errorf("directory %q is not open for reading", id.Path.String())
+	}
+
+	if offset < 0 {
+		return nil, fmt.Errorf("offset %d can't be negative", offset)
+	}
+
+	if id.directory.eos {
+		if offset == id.directory.cursor-1 {
+			return nil, nil // valid end of stream request
+		}
+	}
+
+	if id.directory.eos && offset == id.directory.cursor-1 {
+		return nil, nil // valid end of stream request
+	}
+
+	if offset < id.directory.cursor {
+		return nil, fmt.Errorf("read offset %d is behind current entry %d, seeking backwards in directory streams is not supported", offset, id.directory.cursor)
+	}
+
+	ents := make([]p9.Dirent, 0)
+out:
+	for len(ents) < int(count) {
+		select {
+		case entry, open := <-id.directory.entryChan:
+			if !open {
+				id.directory.eos = true
+				break out
+			}
+			if entry.Err != nil {
+				id.directory = nil
```

Comment:
If we do this, subsequent calls will think the directory is _closed_. Instead, we should stash the error so we can return that on subsequent reads.